src/modm/architecture/interface/can_message.cpp.in

```diff
@@ -10,22 +10,36 @@
 // ----------------------------------------------------------------------------
 
 #include "can_message.hpp"
+#include <inttypes.h>
+#include <string.h>  // strlen
+
+// Converts a single char to its hex value.
+// Potentially duplicated code with modm/driver/can/can_lawicel_formatter.cpp
+uint8_t
+hexchar_to_byte(const char cc)
+{
+	if ((cc >= '0') and (cc <= '9')) {
+		return cc - '0';
+	} else if ((cc >= 'A') and (cc <= 'F')) {
+		return cc - 'A' + 10;
+	} else if ((cc >= 'a') and (cc <= 'f')) {
+		return cc - 'a' + 10;
+	} else {
+		return 0;
+	}
+}
```

Comment:
Technically `strtol(str, NULL, 16)` does this and also exists on avr-libc. However [looking at the implementation](https://github.com/vancegroup-mirrors/avr-libc/blob/master/avr-libc/libc/stdlib/strtol.c), it's fairly expensive.

Perhaps this code should be put into the `modm:math:algorithm` module?

Reply:
Yes, also tried strtol which seemed to be overkill. `algorithm` sound like QR decomposition and advanced stuff, not just stupid bit banging ...

There is `modm:math:utils:misc`🤓

I'll suggest `modm:math:utils:byte_operations`. There already is `modm:math:utils:bit_operations` which includes a `swap` that swaps the nibble of a byte.